### ISSUE TITLE
Using NGLview in Xeus-Python

### ISSUE THREAD

**Issue by kevslinger** [arose/nglview#844]
Perhaps this could have gone in the usage thread #785 

Are there intentions for NGLview to support Xeus-Python? https://github.com/QuantStack/xeus-python I've been experimenting with it to see if we can use it in Clasp https://github.com/clasp-developers/clasp and I noticed nglview has errors in Xeus-Python. I'll explain how to reproduce, but in short, when trying to import nglview, we get the error `ImportError: cannot import name 'HTML'` and `ImportError: cannot import name 'JavaScript'` from IPython.display. We hacked past those errors and found error `Error displaying widget: model not found` with JS errors (looking similar to issue in Jupyterlab https://github.com/jupyterlab/jupyterlab/issues/2791 opened by @hainm). 

- Version report
NGLview: 2.7.0
ipywidgets: 7.5.1

REPRODUCE:
Step 1.) Install Miniconda (required for Xeus)
  a.) Download Miniconda [here](https://docs.conda.io/en/latest/miniconda.html)
  b.) Follow installation instructions [here](https://conda.io/projects/conda/en/latest/user-guide/install/index.html)
  c.) In your shell, run `conda init <your shell name, e.g. zsh or bash>` (You may have to restart your shell after this and before step 2)
Step 2.) Create a Conda environment
  a.) In your Shell, run `conda create -n <shell-name>` (For the rest of this, we will use `xeus-nglview` as shell-name). 
  b.) run `conda activate xeus-nglview`
Step 3.) Install Dependencies
  a.) run `conda install xeus-python -c QuantStack -c conda-forge`
  b.) run `conda install jupyterlab ipywidgets nglview -c conda-forge`
Step 4.) Start Jupyter lab
  a.) run `jupyter lab`
  b.) Create a new notebook with kernel xpython 
  c.) in the kernel, try to import nglview `import nglview`. ERROR: 
```---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
In  [2]:
Line 1:     import nglview

File /Users/kevin/miniconda3/envs/TEST/lib/python3.6/site-packages/nglview/__init__.py, in <module>:
Line 4:     from . import adaptor, datafiles, show, widget

File /Users/kevin/miniconda3/envs/TEST/lib/python3.6/site-packages/nglview/adaptor.py, in <module>:
Line 13:    from .utils.py_utils import FileManager, tempfolder

File /Users/kevin/miniconda3/envs/TEST/lib/python3.6/site-packages/nglview/utils/__init__.py, in <module>:
Line 1:     from . import js_utils, py_utils, widget_utils

File /Users/kevin/miniconda3/envs/TEST/lib/python3.6/site-packages/nglview/utils/js_utils.py, in <module>:
Line 3:     from IPython.display import HTML, Javascript, display

ImportError: cannot import name 'HTML'
---------------------------------------------------------------------------
```
For my experimentation purposes, I don't need HTML features. So I made the following hacks:
1.) In the nglview you installed in your Miniconda environment (`/path/to/miniconda3/envs/xeus-python/lib/python3.6/site-packages/nglview/`), make the following edits:
  a.) In `player.py`, LINE 8: Change `from IPython.display import Javascript, display` to `from IPython.display import display`
  b.) In `nglview/utils/js_utils.py`, LINE 3: remove/comment `from IPython.display import HTML, Javascript, display`, LINES 142, 143, and 158: remove/comment `display(HTML(command))`

Once you have made these edits, you now should be able to `import nglview`. Then, 
step 5.) Try to create widget.
  a.)
 ```
import nglview

view = nglview.show_pdbid("3pqr")  # load "3pqr" from RCSB PDB and display viewer widget

view
```

The output of this cell is `Error displaying widget: model not found`. When we open the inspector, we can see the error

`Unhandled Promise Rejection: Error: Module nglview-js-widgets, semver range 2.7.0 is not registered as a widget module
(anonymous function) — manager.js:251
promiseReactionJob
renderer.js:45`

and also 
`[Error] Error: Could not create a model.`

I'm curious to know if there are plans for NGLview to work in the Xeus-python kernel. Regarding the errors, do we know the cause of the IPython import errors? I find it strange that I do not get those errors in a Python3 ipykernel, but do in a Xpython kernel. Once NGLview is imported, do we know any workarounds/resolutions for the JavaScript issues? Do the hacks I made cause the errors? Any help/guidance/information would be great appreciated.

Thanks so much!

**Comment by hainm** [comment]
hi @kevslinger 

>Are there intentions for NGLview to support Xeus-Python?

I have limited knowledge about python-xeus and this is the first (or may be 2nd) times I heard about it. So there are no intentions for now. Does it work with the ipywidgets itself?
e.g:
```
from ipywidgets import Text
Text("hello")
```

nglview goes with the flow of ipywidgets's ecosystem. If its ecosystem work with xeus-python, we can try to make nglview work with it.

> Regarding the errors, do we know the cause of the IPython import errors?
> I find it strange that I do not get those errors in a Python3 ipykernel, but do in a Xpython kernel. 

My best guess is that xeus-python does not use ipython? But you can safety remove them for your testing.

> Once NGLview is imported, do we know any workarounds/resolutions for the JavaScript issues?

First, make sure the the simple example of ipywidgets (above) works. 
Then make sure to install the `@jupyter-widgets/jupyterlab-manager` and `nglview-js-widgets` for lab-extension. 
We have an example that works for older versions of nglview and jupyterlab but you can adapt it;
https://github.com/arose/nglview/blob/master/devtools/nglview-jupyterlab.sh

> Do the hacks I made cause the errors?
no, they are safe hacks. I suggest to try `nglview` with the traditional notebook in xeus-python first before working with jupyterlab. 

Let me know if I miss anything.
cheers.

**Comment by hainm** [comment]
uhm, I can answer some of my questions:
- xeus-python works with ipywidgets
- xeus-python does not work with `nglview` (after hacking the import in python and some other hacks in Javascript side, see the `xeus` branch here: https://github.com/arose/nglview/tree/xeus)

<img width="1263" alt="Screen Shot 2019-08-21 at 1 32 35 AM" src="https://user-images.githubusercontent.com/4451957/63405110-ab2e2080-c3b3-11e9-8d9d-3e79027dc92f.png">

Basically nglview can show a blank widget but the kernel is disrupted immediately. lol. 

It would be best to raise a question in xeus-python repo to see if it already fully supports 3rd party ipywidgets' extension yet.

**Comment by kevslinger** [comment]
Hi @hainm 
Thank you for your answers! I appreciate the immediate response and the effort you made in the Xeus branch. 

Regarding some of the questions I had, I found an issue related to the IPython import here (https://github.com/QuantStack/xeus-python/issues/106). I guess it is in their backlog. 

I raised the issue in Xeus-Python (https://github.com/QuantStack/xeus-python/issues/124). We'll see what they say and go from there.

Thanks!

**Comment by hainm** [comment]
yeah thanks. I've read about xeus-python yesterday and it has pretty good selling points. I will keep an eye on it.

**Comment by hainm** [comment]
@kevslinger Please try the `xeus` branch again with xeus-python in traditional notebook:

- create a new `xeus` conda env for testing
- install ipywidgets
- install nglview
```
git clone https://github.com/arose/nglview
cd nglview
git checkout xeus
python setup.py install
nglview enable
```

**Comment by kevslinger** [comment]
I get blank with JS errors `TypeError: null is not an object (evaluating 't.filter')` and `TypeError: stage_params is not an Object. (evaluating '"backgroundColor" in stage_params')` I'm assuming this has to do with me not enabling nglview properly? I ran `python setup.py install` and `nglview enable`

![Screen Shot 2019-08-21 at 12 43 08 PM](https://user-images.githubusercontent.com/28906954/63451088-6c2eb800-c411-11e9-9f98-8a8c38f82259.png)

**Comment by hainm** [comment]
did you checkout the `xeus` branch yet?

**Comment by kevslinger** [comment]
yes, checked out Xeus branch before `python setup.py install` and `nglview enable`

**Comment by hainm** [comment]
Do you see this line: https://github.com/arose/nglview/blob/xeus/js/src/widget_ngl.ts#L107

**Comment by kevslinger** [comment]
yes

**Comment by hainm** [comment]
that's odd, I thought this line below would help.

> var stage_params = this.model.get("_ngl_full_stage_parameters") || {};

Can you doublecheck if the JS extension of nglview is installed properly? 
Mine is in: /scr2/nguyen/miniconda3/envs/xeus/share/jupyter/nbextensions/nglview-js-widgets/index.js
Open that file and search for "var stage_params =" to see if that makes sense to you.

**Comment by kevslinger** [comment]
yeah, the file and that line are present in the same location as yours. 

After trying to re-enable everything, I try `nv.demo()` in the notebook, see a blank screen and then the kernel dies. You said this happened to you as well, right? How did you go from there to working?
Thanks

**Comment by hainm** [comment]
I have a fix here: https://github.com/arose/nglview/commit/58cdf991d8b99d067749bb8de498b6e840d3bedb

If you've just git pull the `xeus` branch, the fix should be there. But make sure to doublecheck your code in install folder.

**Comment by kevslinger** [comment]
Still getting the same issue. It must be something I'm doing wrong with either Miniconda or js-enabling or something like that. I get the kernel dying with error
 ```
libc++abi.dylib: terminating with uncaught exception of type pybind11::error_already_set: AttributeError: type object 'kernel._Mock' has no attribute 'showtraceback'

At:
  /Users/kevin/miniconda3/envs/ngl-hack/lib/python3.6/site-packages/ipywidgets/widgets/widget.py(223): m
```

BUT, getting it to work on my machine is more of a "nice-to-have" type of thing. The really important part is that you were able to get nglview working in Xeus-python, so we know that Xeus can support 3rd party widgets. I believe that is the Proof of Concept we needed to start thinking about Xeus-lisp. 

We'll keep an eye on what the Xeus-python folks have to say about this (probably once Sylvain gets back from his time off), and go from there once they come up with a solution they want to go with.

**Comment by hainm** [comment]
> libc++abi.dylib: terminating with uncaught exception of type pybind11::error_already_set: AttributeError: type object 'kernel._Mock' has no attribute 'showtraceback'

are you using macos? if yes, sounds like an issue with xeus-python build itself.
The example works on my Linux machine. If you have one, try with it, if that helps, may be you can open another issue in xeus-python? thanks.

**Comment by hainm** [comment]
> The really important part is that you were able to get nglview working in Xeus-python, so we know that Xeus can support 3rd party widgets. 

For archiving purpose: https://github.com/QuantStack/xeus-python/issues/124#issuecomment-523534740

**Comment by kevslinger** [comment]
> are you using macos? if yes, sounds like an issue with xeus-python build itself.
> The example works on my Linux machine. If you have one, try with it, if that helps, may be you can open another issue in xeus-python? thanks.
yes, I am on Mac. Perhaps that's the issue, good idea.

**Comment by hainm** [comment]
I can reproduce the issue on my mac too.

**Comment by hainm** [comment]
open an issue here: https://github.com/QuantStack/xeus-python/issues/129

**Comment by hainm** [comment]
This can be closed (with "Yes").
